### PR TITLE
DSD-879: Add bgcolor prop to Accordion

### PR DESCRIPTION
**Pull request by oliviawongnyc**
Fixes JIRA tickets [DSD-862](https://jira.nypl.org/browse/DSD-862) and [DSD-879](https://jira.nypl.org/browse/DSD-879).

## This PR does the following:

- Updated accordion styling (border, background color, and padding).
- Added optional key to contentData prop to pass an accordionType that will change an accordion item's background.

## How has this been tested?

- This has been tested locally, in the Test App.

## Accessibility concerns or updates

- No accessibility concerns, but @bigfishdesign13 should take a look at the color choices I made for AccordionTypes.Warning and AccordionTypes.Error and let me know if any changes are needed/how they should look when accordion item is open or hovered over (thanks in advance!).

### Checklist:

- [ ] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.
- [ ] All new and existing tests passed.

### Front End Review:

- [ ] View [the example in Storybook]()

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr919-r6nbmvbnhzzlqcc9sd7fgqnofvjslxtk.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/623e009d628e351b6b7d8937

File: src/theme/components/accordion.ts [hunk 1]
```diff
@@ -1,21 +1,28 @@
...
+    // over is hidden beneath the top border of the following
+    // sibling, so we need to grab and change that border, as well.
+    "& + .chakra-accordion__item": {
+      borderTopColor: "ui.gray.dark",
+    },
+  },
 };
 const buttonStyles = {
   border: "none",
   fontWeight: 500,
```
Comment (bigfishdesign13) [inline src/theme/components/accordion.ts]: There is a `fontWeights` theme object, so this can be set to `medium`.

File: src/theme/components/accordion.ts [hunk 2]
```diff
@@ -1,21 +1,28 @@
 const containerStyles = {
   borderLeftWidth: "1px",
   borderRightWidth: "1px",
-  borderColor: "ui.gray.medium",
   width: "100%",
+  _hover: {
```
Comment (bigfishdesign13) [inline src/theme/components/accordion.ts]: The background color should not change on hover.  The Chakra Accordion must have a default light gray BG color for the hover state, so the styles here should include something to ensure that the BG color that is set for the default state is what shows on hover as well.  Ultimately, only the border color should change on hove.

File: src/components/Accordion/Accordion.stories.mdx [hunk 3]
```diff
@@ -44,23 +45,27 @@ import { getCategory } from "../../utils/componentCategories";
...
 The only way to render an `Accordion` component is to pass in an array of objects
-with `label` and `panel` properties for each accordion item. Note that you can
-pass in a string or DOM elements into the `panel` property in each object. This
-approach is needed because, internally, we deal with the logic to render the
-necessary icon, Chakra components, and styles.
+with `label`, `panel`, and optional `accordionType` properties for each accordion item.
+Note that you can pass in a string or DOM elements into the `panel` property in each
+object. This approach is needed because, internally, we deal with the logic to render
+the necessary icon, Chakra components, and styles. Additionally, the `accordionType` can
+be used to change the accordion button's background color. AccordionTypes.Default is
```
Comment (EdwinGuzman) [inline src/components/Accordion/Accordion.stories.mdx]: I think you can preface the last sentence by saying that the `accordionType` prop takes a value from the `AccordionTypes` enum. Then you can mention each enum value like you already did in the last sentence.

File: src/components/Accordion/Accordion.tsx [hunk 4]
```diff
@@ -46,6 +47,11 @@ const getIcon = (isExpanded = false, index, id) => {
  * combination that is required for the Chakra `Accordion` component.
  */
 const getElementsFromContentData = (data = [], id) => {
+  const colorMap = {
+    default: "ui.white",
```
Comment (EdwinGuzman) [inline src/components/Accordion/Accordion.tsx]: Change the keys to `[AccordionTypes.Default]: "ui.white",` etc.

File: src/theme/components/accordion.ts [hunk 5]
```diff
@@ -1,21 +1,28 @@
...
+    borderLeftWidth: "1px",
+    borderRightWidth: "1px",
```
Comment (EdwinGuzman) [inline src/theme/components/accordion.ts]: These two border values are set above so you don't need them.

File: src/theme/components/accordion.ts [hunk 6]
```diff
@@ -1,21 +1,28 @@
...
+    borderColor: "ui.gray.dark",
```
Comment (EdwinGuzman) [inline src/theme/components/accordion.ts]: Ticket 862 says the border should be `ui.gray.medium` on hover.

File: src/theme/components/accordion.ts [hunk 7]
```diff
@@ -1,21 +1,28 @@
...
+    // sibling, so we need to grab and change that border, as well.
+    "& + .chakra-accordion__item": {
+      borderTopColor: "ui.gray.dark",
+    },
+  },
 };
 const buttonStyles = {
   border: "none",
   fontWeight: 500,
-  _hover: {
```
Comment (EdwinGuzman) [inline src/theme/components/accordion.ts:10]: I think you still need the `_hover` rule but, according to ticket 862, the background should be removed. Chakra seems to have its own default hover bg color. Try making it "transparent" so it removes it rather than "ui.white". I have a feeling that this won't work for the warning and error bg colors though.... So this might be tricky and I think ticket 862 should be updated to reflect the newly added bg colors.

@bigfishdesign13 what do you think?

File: src/theme/components/accordion.ts [hunk 8]
```diff
@@ -1,21 +1,28 @@
...
   width: "100%",
+  _hover: {
+    borderLeftWidth: "1px",
+    borderRightWidth: "1px",
+    borderColor: "ui.gray.dark",
+    // The bottom border of the accordion item being hovered
+    // over is hidden beneath the top border of the following
+    // sibling, so we need to grab and change that border, as well.
+    "& + .chakra-accordion__item": {
+      borderTopColor: "ui.gray.dark",
```
Comment (EdwinGuzman) [inline src/theme/components/accordion.ts]: This should also be `ui.gray.medium`.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
The update is looking good but just needs some minor updates.

File: src/theme/components/accordion.ts (hunk 7 above)
Comment (bigfishdesign13) [reply src/theme/components/accordion.ts:10]: Right, the default BG color from Chakra should not be coming through on hover.  Rather, the initial BG color should remain visible.  In the case of `default`, that would be white.  And so on.

Comment (bigfishdesign13) [reply src/theme/components/accordion.ts:10]: DSD-862 has been updated to clarify the color assignments.

File: src/components/Accordion/Accordion.tsx [hunk 9]
```diff
@@ -10,14 +10,15 @@ import {
 import Icon from "../Icons/Icon";
 import { IconNames, IconSizes } from "../Icons/IconTypes";
 import generateUUID from "../../helpers/generateUUID";
-
+import { AccordionTypes } from "./AccordionTypes";
 export interface AccordionContentDataProps {
+  accordionType?: AccordionTypes;
```
Comment (bigfishdesign13) [inline src/components/Accordion/Accordion.tsx:15]: Instead of making this required, can this be optional with a default value of `AccordionTypes.Default`?

File: src/components/Accordion/Accordion.stories.mdx [hunk 10]
```diff
@@ -44,23 +45,27 @@ import { getCategory } from "../../utils/componentCategories";
 
 The `Accordion` component displays a list of high-level options that can
 expand/collapse to reveal associated sections of content. This component pushes
-existing content on the page down. Each accordion item is self contained but
-when multiple are rendered together, they render in a "FAQ" style. When multiple
+existing content on the page down. Each accordion item is self contained, but
+when multiple are rendered together, they render in an "FAQ" style. When multiple
```
Comment (bigfishdesign13) [inline src/components/Accordion/Accordion.stories.mdx]: I don’t like pigeonholing this component.  How about this:
```
...when the data for multiple accordions is passed in through the `contentData` prop, the `Accordion` components are grouped together and rendered under a single element. When grouped like this, each `Accordion` component still open and close independent from the others in the group.
```

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
The color selections for `Warning` and `Error` look good.  They align with the color treatment used in `StatusBadge`.  I just have a few comments about how the colors are handled for the `hover` and `open` states.

File: src/theme/components/accordion.ts (hunk 6 above)
Comment (oliviawongnyc) [reply src/theme/components/accordion.ts]: The Accordion button border is `ui.gray.medium` on default, so if I add it on hover, there is no change, visually. The Figma file shows `ui.gray.dark` for hover and `ui.gray.medium` for default: https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=10734%3A2520. @bigfishdesign13

File: src/theme/components/accordion.ts (hunk 7 above)
Comment (oliviawongnyc) [reply src/theme/components/accordion.ts:10]: Okay, I moved this logic into the component itself, and it's working. Like so:

```
     <AccordionButton
              id={`${id}-button-${index}`}
              padding={multiplePadding}
              bg={
                content.accordionType
                  ? colorMap[content.accordionType]
                  : colorMap.default
              }
              _expanded={{
                bg:
                  content.accordionType === "default"
                    ? "ui.gray.light-warm"
                    : colorMap[content.accordionType],
              }}
              _hover={{
                bg:
                  content.accordionType === "default"
                    ? "transparent"
                    : colorMap[content.accordionType],
              }}
            >
```

@EdwinGuzman, let me know if you have any issues with this.

File: src/theme/components/accordion.ts (hunk 8 above)
Comment (oliviawongnyc) [reply src/theme/components/accordion.ts]: Same comment as above.

File: src/theme/components/accordion.ts (hunk 2 above)
Comment (oliviawongnyc) [reply src/theme/components/accordion.ts]: @bigfishdesign13, check out my earlier comments on the border color. Maybe the default border color should be something different? Right now, if I set the hover border to `ui.gray.medium` there is no change because the default is already `ui.gray.medium`.

File: src/theme/components/accordion.ts (hunk 7 above)
Comment (EdwinGuzman) [reply src/theme/components/accordion.ts:10]: I don't. Since this is not taking the `__css` prop, separate style props are fine. Make `colorMap[content.accordionType]`  into a variable though so it only does the lookup once.

Comment (oliviawongnyc) [reply src/theme/components/accordion.ts:10]: Will do.

File: src/components/Accordion/Accordion.tsx (hunk 9 above)
Comment (oliviawongnyc) [reply src/components/Accordion/Accordion.tsx:15]: `accordionType` is currently optional, and the component does default to `AccordionTypes.Default` styles if nothing is passed in. Am I understanding this comment correctly or is there something else you'd like to see?

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr919-r6nbmvbnhzzlqcc9sd7fgqnofvjslxtk.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/623e009d628e351b6b7d8937

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr919-r6nbmvbnhzzlqcc9sd7fgqnofvjslxtk.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/623e009d628e351b6b7d8937

File: src/theme/components/accordion.ts (hunk 6 above)
Comment (bigfishdesign13) [reply src/theme/components/accordion.ts]: @oliviawongnyc 
I am sorry for the confusion.  I specified the wrong color value in the Jira ticket.  You have it correct here.  The hover state should use `ui.gray.dark` for border color.

File: src/theme/components/accordion.ts (hunk 8 above)
Comment (bigfishdesign13) [reply src/theme/components/accordion.ts]: Ditto for me.  Same diffident comment as above.

Comment (oliviawongnyc) [reply src/theme/components/accordion.ts]: No worries!! Thanks for clarifying.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr919-r6nbmvbnhzzlqcc9sd7fgqnofvjslxtk.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/623e009d628e351b6b7d8937

File: CHANGELOG.md [hunk 11]
```diff
@@ -12,12 +12,15 @@ Currently, this repo is in Prerelease. When it is released, this project will ad
 
 - Adds a `contentId` prop to the `TemplateAppContainer` component and adds an `id` prop to the `TemplateContent` component. The default value of these `id`s are both set to "mainContent" and it will render as an attribute on the `main` HTML element. This is used as the target for the skip navigation link in consuming applications.
 - Adds an `Accessibility Guide` section to Storybook with a "Skip Navigation" page.
+- Adds an optional key – `accordionType` to the `Accordion`'s `contentData` prop, which allows users to switch the
+  background color of the `Accordion`'s button.
```
Comment (EdwinGuzman) [inline CHANGELOG.md]: Move your _new_ changes to the end of the section. Same with the new line in the `### Changes` section.

Comment (EdwinGuzman) [reply CHANGELOG.md]: Just to keep chronological order of when changes were merged in.

File: src/components/Accordion/Accordion.stories.mdx [hunk 12]
```diff
@@ -96,12 +101,15 @@ const contentData = [
 
 ## Component Props
 
+<ArgsTable story="Accordion with Controls" />
```
Comment (EdwinGuzman) [inline src/components/Accordion/Accordion.stories.mdx]: Why was this moved around?

**Review by EdwinGuzman — APPROVED** [review]
The update looks good but I would change back the order in the Storybook file so the canvas is first before the args table. Approved but wait for @bigfishdesign13 for the final visual review.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr919-r6nbmvbnhzzlqcc9sd7fgqnofvjslxtk.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/623e009d628e351b6b7d8937

File: src/components/Accordion/Accordion.stories.mdx (hunk 12 above)
Comment (oliviawongnyc) [reply src/components/Accordion/Accordion.stories.mdx]: There was a merge conflict in the file when I merged in development and I resolved it by including the current and incoming changes. Because of that, some pieces went into a code section when they were supposed to be part of the text. I moved this to try and make it right again, but I'll have a second look to get it where it needs to be.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr919-r6nbmvbnhzzlqcc9sd7fgqnofvjslxtk.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/623e009d628e351b6b7d8937

**Review by EdwinGuzman — APPROVED** [review]
Just update the snapshot tests but all is good.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr919-r6nbmvbnhzzlqcc9sd7fgqnofvjslxtk.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/623e009d628e351b6b7d8937

File: src/components/Accordion/Accordion.tsx (hunk 9 above)
Comment (bigfishdesign13) [reply src/components/Accordion/Accordion.tsx:15]: Sorry, I mixed up my syntax.  Yes, I now see that prop is optional.  And now that I look even closer at the code, I see am seeing the logic that is defaulting to `default`.

**Review by bigfishdesign13 — APPROVED** [review]
The hover styles look good, so I will approve this to keep things moving forward.  That said, we will need to come back to this component and deal with the double border between the components in the stacked FAQ layout.
